include/xtensor/xstrides.hpp

```diff
@@ -589,6 +596,44 @@ namespace xt
         }
     };
 
+    namespace detail
+    {
+        template <class S, std::size_t dim>
+        inline void check_in_bounds_impl(bool&, const S&)
+        {
+        }
+
+        template <class S, std::size_t dim, class T, class... Args>
+        inline void check_in_bounds_impl(bool& res, const S& shape, T& arg, Args&... args)
+        {
+            if (sizeof...(Args) + 1 > shape.size())
+            {
+                check_in_bounds_impl<S, dim>(res, shape, args...);
+            }
+            else
+            {
+                T n = static_cast<T>(shape[dim]);
+                if (arg < 0 || arg >= n)
+                {
+                    res = false;
+                }
+                check_in_bounds_impl<S, dim + 1>(res, shape, args...);
```

Comment:
I think there might be a problem with passing `res` as reference and never check it in the recursion: the last comparison will be the result whatever the previous comparisons. Also, this prevents logical shortcut. I think `return res && check_in_bounds_impl<S, dim+1>(res, shape, args...);` would be better (same for the `if` case). In that case ,`check_in_bounds_impl` can accept res by value.

Reply:
This is why I pass a reference to `res`. It starts at `true`, and gets `false` if the check on l. 616 fails. If the check does not fail, `res is not modified, so a `false` result gets carried in the recursion. Your implementation, with a `return` `bool`, is more probably more transparent, but also has more boolean comparisons.

Reply:
Ah indeed, my mistake, the first `false` is propagated until the end of the recursion.
Regarding my implementation, actually I think it has less comparisons when the result is false and as much when the result is true; it shortens the implementation and passing `res` as a parameter is actually no long required:

```cpp
return arg >= T(0) && arg < static_cast<T>(shape[dim]) && check_in_bounds_impl<S, dim + 1>(shape, args...);
```

Reply:
I'm convinced, and I've updated accordingly.